### ISSUE TITLE
Change Request: RuleTester should point to the failing test case instead of the failing assertion code

### ISSUE THREAD

**Issue by ST-DDT** [eslint/eslint#19936]
### ESLint version

v9.30.1

### What problem do you want to solve?

The rule tester adds syntactic sugar for defining tests. Since all of the tests are defined as data instead of code, the rule tester doesn't report the location of the test case, instead it reports the location of the assertion (inside the rule tester file).

<img width="869" height="592" alt="Image" src="https://github.com/user-attachments/assets/b54e2a78-f340-4f10-af0d-2431baca0186" />

### What do you think is the correct solution?

Tracking the location where RuleTester.run() is called and then calculate the actual test case definition from it.
This might require searching the file's AST.

````js
run(name, rule, testCases) {
  const traceEntry = new Error().stack[0];
  ...


  for ((testCase, index) of testCases) {
    try {
      verifyAssersions(testCase);
    } catch (error) {
      const testCaseSource = calculateActualSource(traceEntry, index);
      if (testCaseSource) {
        error.stack.shift(testCaseSource);
      }
      throw error;
    }
  }
}
````

<img width="869" height="612" alt="Image" src="https://github.com/user-attachments/assets/a545dd28-f002-4f57-b0a9-0d0efe836c0d" />

If `calculateActualSource` is to complicated (which it likely is), we could start with just pushing the original traceEntry to the stack, so we can jump at least into the correct file.

I did something similar for a different project already:

https://github.com/faker-js/faker/blob/93416f71cf0ddf32dd12c181c862a1b59ef4fd86/test/support/seeded-runs.ts#L337

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by ST-DDT** [comment]
A relatively simple (but partial fix) for this, would be outputting the "invalid" index.

````diff
 if (test.invalid.length > 0) {
 	this.constructor.describe("invalid", () => {
 		test.invalid.forEach((invalid, index) => {
 			this.constructor[invalid.only ? "itOnly" : "it"](
-				sanitize(invalid.name || invalid.code),
+				sanitize(`invalid[${index}]: ${invalid.name || invalid.code}`),
 				() => {
 					try {
 						runHook(invalid, "before");
 						testInvalidTemplate(invalid, index);
 					} finally {
 						runHook(invalid, "after");
 					}
 				},
 			);
 		});
 	});
 }
````

<img width="496" height="120" alt="Image" src="https://github.com/user-attachments/assets/c058311b-e18e-4f1d-bf38-e15c584dd7c3" />

Then you can at least count the invalid test cases to get to the right one.
It would still need something extra for the the invalid[].errors[?] index.

**Comment by ST-DDT** [comment]
Here is an attempt for an error stack extension based variant:

````js
// early during run()
const dummyObject = {};
Error.captureStackTrace(dummyObject, this.run);
const { stack } = dummyObject;
const stackLines = stack.split("\n");
const stackFileSource = stackLines[1];
let testLocations = null;

function estimateTestLocation(type, item, callSpecifier='') { 
  if (testLocations == null) { // Lazy
    const sourceLocation = stackFileSource.replace(/.*\(/, "").replace(/\)$/, "").replace(/:\\/, "\\\\");
    let [sourceFile, sourceLine = '1', sourceColumn = '1'] = sourceLocation.split(':');
    sourceFile = sourceFile.replace(/\\\\/, ':\\'); // Windows workaround
    sourceLine = Number(sourceLine);
    sourceColumn = Number(sourceColumn);
    testLocations = {valid: [], invalid: {}, roots: {}};

    if(existsSync(sourceFile)) {
      const content = readFileSync(sourceFile, "utf8").split("\n").slice(sourceLine - 1);
      content[0] = content[0].substring(sourceColumn - 1);

      const validIndex = content.findIndex(line => line.includes("valid:"));
      const invalidIndex = content.findIndex(line => line.includes("invalid:"));
      testLocations.roots.valid = sourceFile+':'+(sourceLine+validIndex);
      testLocations.roots.invalid = sourceFile+':'+(sourceLine+invalidIndex);

      const validEnd = validIndex<invalidIndex ? invalidIndex : content.length;
      const invalidEnd = validIndex<invalidIndex ? content.length : validIndex;

      const validLines = content.slice(validIndex, validEnd);
      const invalidLines = content.slice(invalidIndex, invalidEnd);

      testLocations.valid = validLines.map((l, i) => l.trimEnd().endsWith(',') ?  sourceFile+':'+(sourceLine+validIndex+i) : null).filter(Boolean);
      testLocations.invalid = invalidLines.map((l, i) => l.trimStart().startsWith("errors:") ? sourceFile+':'+(sourceLine+invalidIndex+i) : null).filter(Boolean);
    }
  }

  switch (type) {
    case "valid":
    case "invalid":
      return `${type}[${item}]${callSpecifier} (${testLocations[type][item]})`;

    case "roots":
      return `${item}${callSpecifier} (${testLocations[type][item]})`;
  }
}

// when running the individual tests:
} catch(assertionError) {
  assertionError.stack += '\n\n        Original Source:\n';
  assertionError.stack += '\n'+ stackFileSource;
  assertionError.stack += '\n    roughly at RuleTester.run.' + estimateTestLocation('roots', 'invalid', `[${index}]`);
  assertionError.stack += '\n    roughly at RuleTester.run.' + estimateTestLocation('invalid', index, `.errors[${i}]`);
  throw assertionError;
}
````

<img width="1055" height="251" alt="Image" src="https://github.com/user-attachments/assets/36bc5791-41bc-4401-8827-04c8bf1c844e" />

It might be better to add the additional items at the top, but at the end it is easier to implement.

**Comment by nzakas** [comment]
Sorry we lost track of this. This seems very helpful. If you're still willing to submit a pull request, we'd love to take a look.

**Comment by ST-DDT** [comment]
No problem, I'm glad my suggestion is considered useful for others.
I created #19976 to add this.

- #19976

**Comment by mdjermanovic** [comment]
Just to clarify, https://github.com/eslint/eslint/pull/19976 has `closes #19936` in the description, but it's just a part of this change and there will be another PR that actually closes this issue?

**Comment by ST-DDT** [comment]
I'm unsure how to actually address the second part without using IO or redesigning the API.

- With IO I can just detect the test call and then guess the actual location
- With an API change like the following, we don't have to guess at all

````js
// Usage
ruleTester.runV2(ruleName, rule, ({valid, invalid}) => {

  valid('Eval()');
  valid({ code: 'Eval()' });

  invalid({ code: "eval()", errors: [{...}] });
});
````

````js
// RuleTester

run(...) // internally calls runV2 to be less/not breaking
runV2(ruleName, rule, fn) {
  const valids = [];

  const valid = (args) => {
    const callLocation = recordCallLocation();
    valids.push({callLocation, args});
  };

  fn({valid, invalid});

  for(const valid in valids) {
    it(valid.args.name, () => {
      try 
        assertValidArgs(...);
        runRule(...);
      } catch(e) {
        patchStacktrace(e, valid.callLocation);
        throw e;
      }
    });
  }
}
````

With that you could even add support for something like `it.only`.

I did something similar for a different project:

https://github.com/faker-js/faker/blob/4cda07f207d84892245c065a99592b350870fcc8/test/support/seeded-runs.ts#L191-L200
https://github.com/faker-js/faker/blob/4cda07f207d84892245c065a99592b350870fcc8/test/modules/string.spec.ts#L10-L24

**Comment by mdjermanovic** [comment]
If we'd go with the proposed API change, then https://github.com/eslint/eslint/pull/19976 wouldn't be necessary?

**Comment by ST-DDT** [comment]
Likely not or at least not in the current form.
That PR is the none breaking least intrucive way to achieve the goal.
It could be extended with the IO variant for better guestimations or replaced by a new API.

**Comment by fasttime** [comment]
I think we could leave the API unchanged and guess the test case location from the code, even if that requires accessing the file system. The approach in #19976 still seems useful for when the location can't be guessed.

To add a new API, we'll probably want to go through the RFC process first in order to discuss the design.

**Comment by ST-DDT** [comment]
Since I already removed the location estimation once.

Should we first review/merge the PR as is and then create a second PR for the actual estimation?

---

Should I create an RFC for that?
Is it worth changing the entire API just for catching rule tester locations? Or are there other improvements to be gained from/to be included in this?

I personally like the freedom individual calls grant, such as creating sections for different options with valid and invalid tests each. E.g. array comma after last element: always, mutliline, never.

Theoretically, we could pick up those "sections" for automated documentation generation, but that would be an entirely separate feature.

**Comment by mdjermanovic** [comment]
> Should we first review/merge the PR as is and then create a second PR for the actual estimation?

To clarify, if we want to keep the current API, https://github.com/eslint/eslint/pull/19976 would be a prerequisite for guessing the exact location of the test case because it provides the file path where the test case is?

**Comment by ST-DDT** [comment]
> To clarify, if we want to keep the current API, [#19976](https://github.com/eslint/eslint/pull/19976) would be a prerequisite for guessing the exact location of the test case because it provides the file path where the test case is?

Only in the sense that it adds the lines to the stacktrace where the exact/guessed location will be added/displayed.

TLDR: That PR currently adds:
` at RuleTester.run.invalid[42].errors[0] (unknown source)`.

**Comment by mdjermanovic** [comment]
> > To clarify, if we want to keep the current API, [#19976](https://github.com/eslint/eslint/pull/19976) would be a prerequisite for guessing the exact location of the test case because it provides the file path where the test case is?
> 
> Only in the sense that it adds the lines to the stacktrace where the exact/guessed location will be added/displayed.
> 
> TLDR: That PR currently adds: ` at RuleTester.run.invalid[42].errors[0] (unknown source)`.

By prerequisite, I meant it provides path to the file and the location within it where `RuleTester#run()` is called, which will be used as data in another PR to read the file and search for the test case in it?

**Comment by ST-DDT** [comment]
The PR as is right now adds the following to the stacktrace:

````ts
    at RuleTester.run.invalid[${index}].error[${errorIndex}] (unknown source)
    at RuleTester.run.invalid[${index}] (unknown source)
    at RuleTester.run.invalid (unknown source)
    at RuleTester.run (${invocationLocation})
````

plus the same for valid.

`invocationLocation` contains the "user" code calling `RuleTester.run`.

**Comment by mdjermanovic** [comment]
@ST-DDT now that https://github.com/eslint/eslint/pull/19976 has been merged, you can submit a PR for the test case location estimation.

**Comment by ST-DDT** [comment]
Sure, I will do that after the xmas days.

**Comment by ST-DDT** [comment]
Done: #20420

- #20420

**Comment by ST-DDT** [comment]
This is now resolved with #20420
